### ISSUE TITLE
keep getting `VM crash or System.exit called?` on CI

### ISSUE THREAD

**Issue by danyaljj** [CogComp/cogcomp-nlp#565]
This has been a recurring exception for the past few weeks: 
```

Results :

Tests run: 5, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] illinois-cogcomp-nlp ............................... SUCCESS [  0.050 s]
[INFO] illinois-core-utilities ............................ SUCCESS [  3.360 s]
[INFO] LBJava-NLP-tools ................................... SUCCESS [  0.360 s]
[INFO] illinois-curator ................................... SUCCESS [  1.614 s]
[INFO] illinois-tokenizer ................................. SUCCESS [  0.467 s]
[INFO] illinois-corpusreaders ............................. SUCCESS [  2.304 s]
[INFO] illinois-edison .................................... SUCCESS [02:04 min]
[INFO] illinois-lemmatizer ................................ SUCCESS [  1.216 s]
[INFO] illinois-pos ....................................... SUCCESS [  3.246 s]
[INFO] illinois-chunker ................................... SUCCESS [ 12.531 s]
[INFO] illinois-ner ....................................... FAILURE [05:07 min]
[INFO] Stanford CoreNLP 3.1.1 ............................. SKIPPED
[INFO] illinois-md ........................................ SKIPPED
[INFO] big-data-utils ..................................... SKIPPED
[INFO] illinois-similarity ................................ SKIPPED
[INFO] illinois-inference ................................. SKIPPED
[INFO] illinois-time ...................................... SKIPPED
[INFO] Illinois Quantifier ................................ SKIPPED
[INFO] illinois-depparse .................................. SKIPPED
[INFO] illinois-comma ..................................... SKIPPED
[INFO] illinois-prep-srl .................................. SKIPPED
[INFO] illinois-verbsense ................................. SKIPPED
[INFO] Illinois NLP Pipeline .............................. SKIPPED
[INFO] External Annotators Commons ........................ SKIPPED
[INFO] ClausIE ............................................ SKIPPED
[INFO] Path-LSTM-SRL-Annotator ............................ SKIPPED
[INFO] Stanford CoreNLP 3.8.0 ............................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 07:36 min
[INFO] Finished at: 2017-10-05T14:17:46-04:00
[INFO] Final Memory: 42M/470M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project illinois-ner: ExecutionException The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[ERROR] Command was /bin/sh -c cd /Users/daniel/ideaProjects/illinois-cogcomp-nlp/ner && /Library/Java/JavaVirtualMachines/jdk1.8.0_72.jdk/Contents/Home/jre/bin/java -Xmx10g -jar /Users/daniel/ideaProjects/illinois-cogcomp-nlp/ner/target/surefire/surefirebooter802195142802041636.jar /Users/daniel/ideaProjects/illinois-cogcomp-nlp/ner/target/surefire/surefire247460483002588515tmp /Users/daniel/ideaProjects/illinois-cogcomp-nlp/ner/target/surefire/surefire_31984716940015435390tmp
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :illinois-ner
```

In fact it fails on my machine as well (not just CI).

**Comment by danyaljj** [comment]
it might just be a memory issue. Increase the memory in #564 and so far it's acting fine ....

**Comment by cowchipkid** [comment]
Maybe OOM-killer has been up to it's evil deeds again? This think force kills apps when it sees the OS running low on memory. So if we are allocating too much memory to the JVM, we might be draining the OS... This would also be erratic, it would depend on what the OS was doing.

**Comment by danyaljj** [comment]
While increasing the memory limit helped it, but the issue still happens.

**Comment by danyaljj** [comment]
No luck. 
@cowchipkid See the build history here: 
http://morgoth.cs.illinois.edu:5800/viewType.html?buildTypeId=CogcompNlp_Build&tab=buildTypeHistoryList

As you can in the changes I tried increasing the memory too. Any thoughts? Do you see the issue locally? Because it happens to me, and it's 70% of the cases ner, and the rest ner being called from the pipeline.

**Comment by cowchipkid** [comment]
I just saw this issue on my machine, while running NER tests. I don't understand this at all, don't think it's OOM-killer issue, OOM issue within the JVM should simply throw an exception and move on (well, if it can), there are no core files, and the sys log files don't say anything about a fault (that I can see). I will try to find some time to play with this, think I might work next Monday and Tuesday rather than Thurs and Friday, that way I can look at this in more detail.

**Comment by danyaljj** [comment]
Thanks @cowchipkid would be great if you give it a high priority. Basically our progress in the codebase is hampered by this, since CI always fails.

**Comment by cowchipkid** [comment]
@danyaljj  Yes, I would have given this it's due before, but the SRL Ontonotes 5 parsing is also a high priority. I think I have that in hand now. This one is going to be tricky.

**Comment by cowchipkid** [comment]
@danyaljj @mssammon I have built cogcomp-nlp dozens of times, different permutations and combinations, excluding subproject here and there, changing heap size and so on. It's very erratic, and inconsistent. I have built this thing most successfully on a machine where the 1.6 gazetteers are again the older and smaller ones. I thought we had fixed this beyond the 1.5 version, but it seems to be back. With the bigger gazetteers, this problem seemed to be much easier to reproduce. With the smaller ones, it's fairly infrequent. I am going to keep working this, but these are my current hypotheses:

Surefire, or something maven related could be calling System.exit(). Since this happens all over the place, I don't think it's a System.exit we call or we would see it in the same place every time. Since the issue triggers more commonly during pipeline and ner testing, it may be related to memory size. Perhaps it has to do with the cleanup time required for these applications to shut down cleanly, maybe surefire has some kind of timer that going off.

I have seen no evidence of the systems OOM killer triggering this, nothing in the logs to confirm that.

Not JVM version related, I have seen this on Sun and OpenJDK(although I would recommend we quit using OpenJDK).

If there is a hs_err file anywhere, I have not seen it, so it's not likely this is a segfault thrown by the VM. Maybe surefire puts this somewhere else...

I am out of Roth time for this week, I will keep picking at this scab evenings see if I can't find other clues. Current clues are the lack of OOM - killer logging, hs_err files indicating seg faults, and exacerbated occurrences with higher memory usage.
